### ISSUE TITLE
RSC shows poor performance OfficeHome

### ISSUE THREAD

**Issue by lopezpaz** [facebookresearch/DomainBed#20]
@SirRob1997: Preliminar runs show a very poor performance (39%) at OfficeHome. The paper reports 63%. Could you investigate how to get a higher number in this dataset with default hyper-parameters?

_Originally posted by @lopezpaz in https://github.com/facebookresearch/DomainBed/pull/12#issuecomment-697476956_

**Comment by lopezpaz** [comment]
Reimplementing RSC following Featurizer/Classifier guidelines for the rest of the algorithms.

**Comment by SirRob1997** [comment]
Yeah the algorithm has some flaws, also in their original implementation it seems like the performance varies a lot (also on PACS). Further, they also adapted their implementation with downsampling of middle layers for ResNet which is not mentioned in the paper. I‘m currently running experiments to get performance up, let’s see   how these turn out.

**Comment by lopezpaz** [comment]
Minimal implementation, will investigate from here:

```python
class RSC(ERM):
    def __init__(self, input_shape, num_classes, num_domains, hparams):
        super(RSC, self).__init__(input_shape, num_classes, num_domains,
                                   hparams)
        self.percentile = (1 - hparams['rsc_f_drop_factor']) * 100
        self.num_classes = num_classes

    def update(self, minibatches):
        # inputs
        all_x = torch.cat([x for x,y in minibatches])
        # labels
        all_y = torch.cat([y for _,y in minibatches])
        # one-hot labels
        all_o = torch.nn.functional.one_hot(all_y, self.num_classes)
        # features
        all_f = self.featurizer(all_x) 
        # predictions
        all_p = self.classifier(all_f)

        # Equation (1): compute gradients with respect to representation
        all_g = autograd.grad((all_p * all_o).sum(), all_f)[0]
       
        # Equation (2): compute top-gradient-percentile mask 
        percentiles = np.percentile(all_g, self.percentile, axis=1)
        percentiles = torch.Tensor(percentiles)
        percentiles = percentiles.unsqueeze(1).repeat(1, all_g.size(1))
        mask = all_g.lt(percentiles).float()

        # Equation (3): mute top-gradient-percentile activations
        all_muted = all_f * mask

        # Equation (4): compute muted predictions
        muted_predictions = self.classifier(all_muted)

        # Equation (5): update
        loss = F.cross_entropy(muted_predictions, all_y)
        self.optimizer.zero_grad()
        loss.backward() 
        self.optimizer.step()

        return {'loss': loss.item()}
```

**Comment by SirRob1997** [comment]
Will be interesting to see how this will change the performance as this operates on features after average pooling instead of the suggested conv4 output. One thing which might be the cause of the degrading performance is that I was missing the following line after [this line](https://github.com/facebookresearch/DomainBed/blob/master/domainbed/algorithms.py#L884).

`drop_indeces = drop_indeces.nonzero()[:,0]`

This recovers the indices and without it the masking gets applied falsely.

**Comment by lopezpaz** [comment]
The above minimal implementation achieves:

```
-------- Averages, model selection method: training-domain validation set
Algorithm                  ColoredMNIST               OfficeHome                 PACS                       RotatedMNIST               TerraIncognita             VLCS                      
RSC                        51.6 +/- 0.1               65.7 +/- 0.1               84.4 +/- 0.3               97.9 +/- 0.1               47.1 +/- 1.0               77.0 +/- 0.8
```

**Comment by SirRob1997** [comment]
My implementation with the fix for the batching got these results:

![image](https://user-images.githubusercontent.com/20804972/94273677-ff0b2180-ff44-11ea-9ea1-f5dde9b46e8a.png)

**Comment by SirRob1997** [comment]
Looks quite good actually, didn't expect that. Applying the mask only for top-p percentile samples in the batch should boost performance quite a bit.

**Comment by lopezpaz** [comment]
I see, I had missed this option by directly implementing Algorithm 1 as described in Section 3.1. of the paper. Will do!

**Comment by SirRob1997** [comment]
Algorithm 1 misses some of the key aspects of their implementation. Take a look at Section 3.3 where they introduce Spatial-Wise and Channel-Wise RSC and the stuff for applying them only for certain samples inside each batch. My initial implementation should have these features (with the small fix mentioned above) but can probably be written a lot cleaner.

**Comment by lopezpaz** [comment]
The minimal version above shows good performance. Next, I will try this version including batch dropping:

```python
class RSC(ERM):
    def __init__(self, input_shape, num_classes, num_domains, hparams):
        super(RSC, self).__init__(input_shape, num_classes, num_domains,
                                   hparams)
        self.drop_f = (1 - hparams['rsc_f_drop_factor']) * 100
        self.drop_b = (1 - hparams['rsc_b_drop_factor']) * 100
        self.num_classes = num_classes

    def update(self, minibatches):
        # inputs
        all_x = torch.cat([x for x, y in minibatches])
        # labels
        all_y = torch.cat([y for _, y in minibatches])
        # one-hot labels
        all_o = torch.nn.functional.one_hot(all_y, self.num_classes)
        # features
        all_f = self.featurizer(all_x)
        # predictions
        all_p = self.classifier(all_f)

        # Equation (1): compute gradients with respect to representation
        all_g = autograd.grad((all_p * all_o).sum(), all_f)[0]

        # Equation (2): compute top-gradient-percentile mask
        percentiles = np.percentile(all_g.cpu(), self.drop_f, axis=1)
        percentiles = torch.Tensor(percentiles)
        percentiles = percentiles.unsqueeze(1).repeat(1, all_g.size(1))
        mask = all_g.lt(percentiles.cuda()).float()

        # Equation (3): mute top-gradient-percentile activations
        all_f_muted = all_f * mask

        # Equation (4): compute muted predictions
        all_p_muted = self.classifier(all_f_muted)

        # Section 3.3: Batch Percentage
        all_s = F.softmax(all_p, dim=1)
        all_s_muted = F.softmax(all_p_muted, dim=1)
        changes = (all_s * all_o).sum(1) - (all_s_muted * all_o).sum(1)
        percentile = np.percentile(changes.detach().cpu(), self.drop_b)
        mask *= changes.lt(percentile).float().view(-1, 1)

        # Equations (3) and (4) again, this time mutting over examples
        all_p_muted_again = self.classifier(all_f * mask)

        # Equation (5): update
        loss = F.cross_entropy(all_p_muted_again, all_y)
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()

        return {'loss': loss.item()}
```
I haven't implemented spatial dropping yet, because currently all algorithms in DomainBed perform regularization in a `(n_examples, n_features)` bottleneck followed by a linear classifier.

Thank you for your help!
